pkg/cri/resource-manager/policy/builtin/balloons/flags.go

```diff
@@ -0,0 +1,254 @@
+	// attempts to fill a balloon instance of chosen type turn out
+	// to be impossible but do not raise error, the next method in
+	// the AssignChain will be tried out.
+	AssignChain []AssignMethod `json:"AssignChain,omitempty"`
+	// FillChain chooses a balloon instance to fill with new
+	// container. The first successful method in the list is
+	// applied.
+	FillChain []FillMethod `json:"FillChain,omitempty"`
+	// BallonDefs contains balloon type definitions.
+	BalloonDefs []*BalloonDef `json:"BalloonTypes,omitempty"`
+}
+
+// BalloonDef contains a balloon definition.
+type BalloonDef struct {
+	// Name of the balloon definition.
+	Name string `json:"Name"`
+	// Namespaces control which namespaces are assigned into
+	// balloon instances from this definition. This is used by
+	// namespace assign methods.
+	Namespaces []string `json:"Namespaces",omitempty`
+	// MaxCpus specifies the maximum number of CPUs exclusively
+	// usable by containers in a balloon. Balloon size will not be
+	// inflated larger than MaxCpus.
+	MaxCpus int `json:"MaxCPUs"`
+	// MinCpus specifies the minimum number of CPUs exclusively
+	// usable by containers in a balloon. When new balloon is created,
+	// this will be the number of CPUs reserved for it even if a container
+	// would request less.
+	MinCpus int `json:"MinCPUs"`
+	// AllocatorPriority (0: High, 1: Normal, 2: Low, 3: None)
+	// This parameter is passed to CPU allocator when creating or
+	// resizing a balloon. At init, balloons with highest priority
+	// CPUs are allocated first.
+	AllocatorPriority cpuallocator.CPUPriority `json:"AllocatorPriority"`
+	// CpuConf controls how CPUs of a balloon are (re)configured
+	// whenever balloon is created, inflated or deflated.
+	CpuConf map[string]string `json:"CPUConf"`
+	// MinBalloons is the number of balloon instances that always
+	// exist even if they would become empty. At init this number
+	// of instances will be created before assigning any
+	// containers.
+	MinBalloons int `json:"MinBalloons"`
+	// FillChain control how multi-instance balloons are filled.
+	// Overrides common FillChain.
+	FillChain []FillMethod `json:"FillChain"`
+}
+
+// AssignMethod specifies how to select balloon type for a container
+type AssignMethod int
+
+// FillMethod specifies the order in which balloon instances should be filled.
+type FillMethod int
+
+const (
+	AssignUnspecified AssignMethod = iota
+	// AssignNamespace: assign a container to a balloon type whose
+	// Namespaces list contains container's namespace.
+	AssignNamespace
+	// AssignNamespaceMust: refuse to run a container if its
+	// namespace is not in any balloon's namespaces list.
+	AssignNamespaceMust
+	// AssignAnnotation: container's balloon annotation matches
+	// balloon Name.
+	AssignAnnotation
+	// AssignAnnotationMust: refuse to run a container without a
+	// balloon matching the annotation.
+	AssignAnnotationMust
+	// Assign kube-system containers into the reserved balloon.
+	AssignSystemToReserved
```

Comment:
I can't help the feeling that it is just wrong (or at least a huge source for unintentional misconfiguration) for having such a thing to exist.

Reply:
AssignChain completely removed.